src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py

```diff
@@ -0,0 +1,580 @@
+        """
+        return OBSPruningParamsScorer(
+            params=params,
+            num_grads=self._num_grads,
+            damp=self._damp,
+            fisher_block_size=self._fisher_block_size,
+            mask_type=self._mask_type,
+        )
+
+    def _collect_grad_samples(
+        self,
+        module: Module,
+        grad_sampler: GradSampler,
+    ):
+        if not isinstance(grad_sampler, GradSampler):
+            raise ValueError(
+                "One-shot OBS pruning requires a GradSampler object given by the "
+                f"grad_sampler kwarg. Given an object of type {type(grad_sampler)}"
+            )
+
+        is_training = module.training
+        _LOGGER.debug("Setting the model in the eval mode")
+        module.eval()
+
+        _LOGGER.debug(f"Starting to collect {self._num_grads} grads with GradSampler")
+        for _ in grad_sampler.iter_module_backwards(module, self._num_grads):
+            self._module_masks.pre_optim_step_update()
+
+        if is_training:
+            _LOGGER.debug("Setting the model back to the train mode")
+            module.train()
+
+    def _validate(self):
+        if not hasattr(torch.linalg, "solve"):
+            raise RuntimeError(
+                f"torch>=1.9 required to use {self.__class__.__name__} "
+                f"found {torch.__version__}"
+            )
+
+        if isinstance(self._damp, str):  # to support 'damp: 1e-7' in the recipe
+            self._damp = float(self._damp)
+
+        if self._mask_type not in self._supported_masks:
+            raise ValueError(f"{self._mask_type} mask_type not supported")
+
+        if self._mask_type == "block4" and self._fisher_block_size % 4 != 0:
+            raise ValueError(
+                "fisher_block_size must be divisible by 4 for block4 pruning"
+            )
+
+
+class OBSPruningParamsScorer(PruningParamsGradScorer):
+    """
+    Scores parameters using the equations introduced in the Optimal BERT Surgeon
+    to solve for the optimal weight update in the Optimal Brain Surgeon (OBS)
+    framework. Implements unstructured and semi-structured (block4) scoring and
+    pruning.
+
+    :param params: list of model Parameters to track and score
+    :param num_grads: number of gradients used to calculate the Fisher approximation
+    :param damp: dampening factor, default is 1e-7
+    :param fisher_block_size: size of blocks along the main diagonal of the Fisher
+        approximation, default is 50
+    """
+
+    def __init__(
+        self,
+        params: List[Parameter],
+        num_grads: int,
+        damp: float,
+        fisher_block_size: int,
+        mask_type: str,
+    ):
+        super().__init__(params)
+        self._num_grads = num_grads
+        self._damp = damp
+        self._fisher_block_size = fisher_block_size
+        self._mask_type = mask_type
+
+        self._finvs = None  # type: List[EmpiricalBlockFisherInverse]
```

Comment:
Why can't we directly attach the `type` to the variable?

Reply:
do you mean like `self._finvs: List[EmpiricalBlockFisherInverse]` or something else?

Reply:
the `type:` in-line comment is used throughout sparseml. resolving comment